core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1531 @@
+      if (!topicLevelThrottles.flatMap(throttle => Option(configs.remove(throttle))).isEmpty) {
+        adminZkClient.changeTopicConfig(topic, configs)
+      }
+    }
+  }
+
+  /**
+   * The entry point for the --generate command.
+   *
+   * @param adminClient           The AdminClient to use.
+   * @param reassignmentJson      The JSON string to use for the topics to reassign.
+   * @param brokerListString      The comma-separated string of broker IDs to use.
+   * @param enableRackAwareness   True if rack-awareness should be enabled.
+   *
+   * @return                      A tuple containing the proposed assignment and the
+   *                              current assignment.
+   */
+  def generateAssignment(adminClient: Admin,
+                         reassignmentJson: String,
+                         brokerListString: String,
+                         enableRackAwareness: Boolean)
+                         : (Map[TopicPartition, Seq[Int]], Map[TopicPartition, Seq[Int]]) = {
+    val (brokersToReassign, topicsToReassign) =
+      parseGenerateAssignmentArgs(reassignmentJson, brokerListString)
+    val currentAssignments = getReplicaAssignmentForTopics(adminClient, topicsToReassign)
+    val brokerMetadatas = if (enableRackAwareness) {
+      getBrokerRackInformation(adminClient, brokersToReassign)
+    } else {
+      brokersToReassign.map(new BrokerMetadata(_, None))
+    }
+    val proposedAssignments = calculateAssignment(currentAssignments, brokerMetadatas)
+    println("Current partition replica assignment\n%s\n".
+      format(formatAsReassignmentJson(currentAssignments, Map.empty)))
+    println("Proposed partition reassignment configuration\n%s".
+      format(formatAsReassignmentJson(proposedAssignments, Map.empty)))
+    (proposedAssignments, currentAssignments)
+  }
+
+  /**
+   * The legacy entry point for the --generate command.
+   *
+   * @param zkClient              The ZooKeeper client to use.
+   * @param reassignmentJson      The JSON string to use for the topics to reassign.
+   * @param brokerListString      The comma-separated string of broker IDs to use.
+   * @param enableRackAwareness   True if rack-awareness should be enabled.
+   *
+   * @return                      A tuple containing the proposed assignment and the
+   *                              current assignment.
+   */
+  def generateAssignment(zkClient: KafkaZkClient,
+                         reassignmentJson: String,
+                         brokerListString: String,
+                         enableRackAwareness: Boolean)
+                         : (Map[TopicPartition, Seq[Int]], Map[TopicPartition, Seq[Int]]) = {
+    val (brokersToReassign, topicsToReassign) =
+      parseGenerateAssignmentArgs(reassignmentJson, brokerListString)
+    val currentAssignments = zkClient.getReplicaAssignmentForTopics(topicsToReassign.toSet)
+    val brokerMetadatas = if (enableRackAwareness) {
```

Comment:
In the [old code](https://github.com/apache/kafka/pull/8244/files#diff-15a36057cf9f21154f103a55757a7b59L186), we always called `AdminZkClient#getBrokerMetadatas` regardless of the rack-aware mode. The only difference I can see for the `RackAwareMode.Disabled` case is that the old code would always filter down to the set of brokers register in ZK at the time of the call

```scala
def getBrokerMetadatas(rackAwareMode: RackAwareMode = RackAwareMode.Enforced,
                         brokerList: Option[Seq[Int]] = None): Seq[BrokerMetadata] = {
    val allBrokers = zkClient.getAllBrokersInCluster
    val brokers = brokerList.map(brokerIds => allBrokers.filter(b => brokerIds.contains(b.id))).getOrElse(allBrokers)
    ...
    case RackAwareMode.Disabled => brokers.map(broker => BrokerMetadata(broker.id, None))
```

Does this mean that the new code will generate an assignment for a broker that is currently offline, whereas before it would not?

Reply:
Good call.  Yeah, we should probably continue the old behavior of ignoring brokers that aren't up at the time of generating the plan, since there's no specific reason to change it.  Fixed.